lake/branch.go

```diff
@@ -0,0 +1,441 @@
+package lake
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/brimdata/zed/expr/extent"
+	"github.com/brimdata/zed/lake/commit"
+	"github.com/brimdata/zed/lake/index"
+	"github.com/brimdata/zed/lake/journal"
+	"github.com/brimdata/zed/lake/segment"
+	"github.com/brimdata/zed/order"
+	"github.com/brimdata/zed/pkg/nano"
+	"github.com/brimdata/zed/pkg/storage"
+	"github.com/brimdata/zed/zio"
+	"github.com/brimdata/zed/zio/zngio"
+	"github.com/brimdata/zed/zng"
+	"github.com/brimdata/zed/zson"
+	"github.com/segmentio/ksuid"
+)
+
+type BranchConfig struct {
+	Name   string      `zng:"name"`
+	ID     ksuid.KSUID `zng:"id"`
+	Parent ksuid.KSUID `zng:"parent"`
+}
+
+type Branch struct {
+	BranchConfig
+	pool   *Pool
+	engine storage.Engine
+	log    *commit.Log
+	base   commit.View
+}
+
+func newBranchConfig(name string, parent ksuid.KSUID) *BranchConfig {
+	//XXX need to take parent args and check for a cycle.
+	// also check that parent journal ID exists in parent.
+	return &BranchConfig{
+		Name:   name,
+		ID:     ksuid.New(),
```

Comment:
I'm kind of wondering why even bother having an ID for branches. I was moving through the zedlake fs and found it harder to follow what was going on with the ids for branches. I get why we did this for pool names- we didn't want to break links when pools were renamed. This feels like less of a problem with branches.

Reply:
Good point but as we discussed... leaving fo now

Reply:
They will go away in branches v2.